src/Microsoft.ServiceHub.Framework/AsyncNamedPipeClientStream.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO.Pipes;
+using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
+using System.Security.Principal;
+using Microsoft.Win32.SafeHandles;
+
+namespace Microsoft.ServiceHub.Framework
+{
+	/// <summary>
+	/// Named pipe client that avoids TimeoutExceptions and burning CPU.
+	/// </summary>
+	[SupportedOSPlatform("windows")]
+	internal class AsyncNamedPipeClientStream : PipeStream
+	{
+		private const int SECURITYSQOSPRESENT = 0x00100000;
+		private const int GENERICREAD = unchecked((int)0x80000000);
+		private const int GENERICWRITE = 0x40000000;
+		private readonly string pipePath;
+		private readonly TokenImpersonationLevel impersonationLevel;
+		private readonly PipeOptions pipeOptions;
+		private readonly PipeDirection direction;
+		private readonly int access;
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="AsyncNamedPipeClientStream"/> class.
+		/// </summary>
+		/// <param name="serverName">Pipe server name.</param>
+		/// <param name="pipeName">Pipe name.</param>
+		/// <param name="direction">Communication direction.</param>
+		/// <param name="options">Pipe options.</param>
+		/// <param name="impersonationLevel">Impersonation level.</param>
+		internal AsyncNamedPipeClientStream(
+			string serverName,
+			string pipeName,
+			PipeDirection direction,
+			PipeOptions options,
+			TokenImpersonationLevel impersonationLevel = TokenImpersonationLevel.None)
+			: base(direction, 4096)
+		{
+			Requires.NotNullOrEmpty(serverName, nameof(serverName));
+			Requires.NotNullOrEmpty(pipeName, nameof(pipeName));
+
+			this.pipePath = $@"\\{serverName}\pipe\{pipeName}";
+			this.direction = direction;
+			this.pipeOptions = options;
+			this.impersonationLevel = impersonationLevel;
+			this.access = this.GetAccess();
+		}
+
+		/// <summary>
+		/// Connects pipe client to server.
+		/// </summary>
+		/// <param name="cancellationToken">Cancellation token.</param>
+		/// <param name="maxRetries">Maximum number of retries.</param>
+		/// <param name="retryDelayMs">Milliseconds delay between retries.</param>
+		/// <returns>A task representing the establishment of the client connection.</returns>
+		/// <exception cref="TimeoutException">Thrown if no connection can be established.</exception>
+		public async Task ConnectAsync(
```

Comment:
💅🏻 avoid using `public` on members of an `internal` type unless it implicitly implements an interface or overrides a public member. That way, it's very clear to everyone glancing at this (like a code reviewer) that this method does _not_ contribute to the public API, which generally calls for extra review.